pyfifotax/data_structures_dataframe.py

```diff
@@ -0,0 +1,453 @@
+            )
+
+        sale_price = pd.to_numeric(sale_price.strip("$").replace(",", ""))
+
+        return SellOrderRow(
+            date,
+            symbol,
+            quantity,
+            sale_price,
+            fees,
+            "USD",
+            "Automated Schwab Import (JSON)",
+        )
+
+    @staticmethod
+    def default_dict() -> dict:
+        return SellOrderRow(
+            datetime(1, 1, 1), "", np.float64(0), np.float64(0), np.float64(0), "", ""
+        ).to_dict()
+
+    @staticmethod
+    def from_df_row(row: Series) -> SellOrderRow:
+        return SellOrderRow(
+            row.date,
+            row.symbol,
+            row.quantity,
+            row.sell_price,
+            row.fees,
+            row.currency,
+            row.comment,
+        )
+
+
+@dataclass
+class BuyOrderRow(DataFrameRow):
+    date: datetime
+    symbol: str
+    quantity: np.float64
+    buy_price: np.float64
+    fees: np.float64
+    currency: str
+    comment: str
+
+    @staticmethod
+    def default_dict() -> dict:
+        return BuyOrderRow(
+            datetime(1, 1, 1), "", np.float64(0), np.float64(0), np.float64(0), "", ""
+        ).to_dict()
+
+    @staticmethod
+    def from_df_row(row: Series) -> BuyOrderRow:
+        return BuyOrderRow(
+            row.date,
+            row.symbol,
+            row.quantity,
+            row.buy_price,
+            row.fees,
+            row.currency,
+            row.comment,
+        )
+
+
+@dataclass
+class CurrencyConversionRow(DataFrameRow):
+    date: datetime
+    foreign_amount: np.float64
+    source_fees: np.float64
+    source_currency: str
+    target_currency: str
+    comment: str
+
+    @staticmethod
+    def from_schwab_json(json_dict: dict) -> CurrencyConversionRow:
+        date = datetime.strptime(json_dict["Date"], "%m/%d/%Y")
+        fees = pd.to_numeric(
+            json_dict["FeesAndCommissions"].strip("-$").replace(",", "")
+        )
+        foreign_amount = pd.to_numeric(json_dict["Amount"].strip("-$").replace(",", ""))
+        return CurrencyConversionRow(
+            date,
+            foreign_amount,
+            fees,
+            "USD",
+            "EUR",
+            "Automated Schwab Import (JSON, Currency Conversion from Wire Transfer, check correctness!)",
+        )
+
+    @staticmethod
+    def default_dict() -> dict:
+        return CurrencyConversionRow(
+            datetime(1, 1, 1), np.float64(0), np.float64(0), "", "", ""
+        ).to_dict()
+
+    @staticmethod
+    def from_df_row(row: Series) -> CurrencyConversionRow:
+        return CurrencyConversionRow(
+            row.date,
+            row.foreign_amount,
+            row.source_fees,
+            row.source_currency,
+            row.target_currency,
+            row.comment,
+        )
+
+
+@dataclass
+class CurrencyMovementRow(DataFrameRow):
+    date: datetime
+    buy_date: datetime
```

Comment:
This is needed to continue keeping tracking of the FOREX FIFO queue. When you transfer foreign currencies, you still need to remember the acquisition date, yes.
I'll agree with buy_date not being the best name. acquisition indeed might be better. Technically, it would be something like `date_of_initial_exchange` or something. Even receiving USD from something (see how RSUs/ESPPs are treated), technically corresponds to receiving a EUR balance which is then converted to either shares or foreign currency.

Reply:
`date` thus is the "date" of either withdrawal or deposit, i.e. follows the main logic of all other entries that `date` corresponds to the "date" of the event. 

For withdrawals, only "date" is important and corresponds to the date of money being pulled out an account following FIFO (I am actually not really sure how accounts or our tax system handle that but I would assume that implicitly FIFO is followed).

For deposits, we need the additional date of the initial exchange to calculate a potential gain or loss if we end up selling it.